[margin-trim] block-start margins for grid items in horizontal writing-mode should be reflected in computed style.

sammygill opened the pull request:
#### c87ee9281557b041309d82cc2a04fff613444fc9
<pre>
[margin-trim] block-start margins for grid items in horizontal writing-mode should be reflected in computed style.
<a href="https://bugs.webkit.org/show_bug.cgi?id=253716">https://bugs.webkit.org/show_bug.cgi?id=253716</a>
rdar://106559103

Reviewed by Alan Baradlay.

When the block-start margins for grid items that are adjacent to the
block-start edge of the grid, We need to set the rare data bits for
the item to indicate that the margin has been trimmed. We can then check
if this bit is set within ComputedStyleExtractor to get the trimmed
value for this margin by directly querying the renderer&apos;s m_marginBox.
In a horizontal writing mode this bit would correspond to the &quot;top,&quot;
margin value.

For block-start margins, the trimming occurs when a renderer goes
through layout and calls RenderBox::constrainBlockMarginInAvailableSpaceOrTrim,
which is called from RenderBox::computeBlockDirectionMargins. This
function consults the layout system, which is grid in this case, via
shouldTrimChildMargin to see if it should trim its margin. If this
returns true, then the renderer will not compute the margin and will
return 0_lu. We can use this opportunity to set the rare data for the
renderer to indicate that this margin has been trimmed.

Inside of ComputedStyleExtractor, we need to update
rendererCanHaveTrimmedMargin so that this returns true for grid items.
This will allow us to do 2 things:
1.) Have isLayoutDependent return true for grid items with margin-trim
set on the grid so that layout is forced when checking the &quot;top,&quot; margin
for grid items
2.) Allow us to call box-&gt;hasTrimmedMargin when computing the margin
value (since that call is guarded by a call to rendererCanHaveTrimmedMargin)

Similarly, there are a couple of other extra pieces of code in RenderBox
(particularly in physicalToFlowRelativeDirectionMapping and hasTrimmedMargin)
that should now be used with grid items.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/grid-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/grid-block-start.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::rendererCanHaveTrimmedMargin):
(WebCore::isLayoutDependent):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::physicalToFlowRelativeDirectionMapping const):
(WebCore::RenderBox::hasTrimmedMargin const):
(WebCore::RenderBox::constrainBlockMarginInAvailableSpaceOrTrim const):

Canonical link: <a href="https://commits.webkit.org/262967@main">https://commits.webkit.org/262967@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f41c7f021fae7b4f979f523e3c610cbfbb26054f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3195 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4547 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3495 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3103 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3256 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3222 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2725 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3166 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3486 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2812 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4353 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/984 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2787 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2631 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2758 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2838 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4104 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3210 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2562 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2796 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2789 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/772 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2787 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1202,6 +1206,20 @@ GridTrackSizingDirection RenderGrid::autoPlacementMinorAxisDirection() const
 
+void RenderGrid::trimItemMarginsAdjacentToGridEdge(MarginTrimType marginTrimType)
+{
+    if (marginTrimType != MarginTrimType::BlockStart) {
```
This statement is redundent. `setTrimmedMarginForChild` should handle this case.

Let's just remove it.

sammygill replied:
After all the flexbox patches I think this is no longer redundant. I think the early return would have been useful in either case because the logic below only applies to the block-start margin  and would be incorrect for other values of `MarginTrimType` on non-debug builds

stwrt commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2304,14 +2304,19 @@ static bool rendererCanHaveTrimmedMargin(const RenderBox& renderer, std::optiona
     }
+    if (containingBlock->isRenderGrid()) {
+        if (!marginTrimType)
+            return containingBlock->style().marginTrim().containsAny( { MarginTrimType::BlockStart });
```
In other places you are using `.contains`. I am not sure if `containsAny` is getting us anything since there is only one element.

sammygill replied:
Yeah that's fair since in this patch block-start is the only valid value to consider in this context. Went ahead and changed it

stwrt reviewed (changes requested):
A few things I noticed.

stwrt reviewed (changes requested):
A few things I noticed.

alanbaradlay commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1202,6 +1206,20 @@ GridTrackSizingDirection RenderGrid::autoPlacementMinorAxisDirection() const
+    GridIterator iterator(currentGrid(), ForRows, 0);
+    while (auto* gridItem = iterator.nextGridItem()) {
+        if (shouldTrimChildMargin(marginTrimType, *gridItem))
+            setTrimmedMarginForChild(*gridItem, marginTrimType);
```
I was under the impression that this bit is set when the margin is being trimmed (like in RenderFlexibleBox::trimMainAxisMarginStart). How's grid different in this regard? Are we re-trimming here by calling setTrimmedMarginForChild?

sammygill replied:
Flex is slightly different because of some additional consideration that is needed in the multi-line flexbox scenario. `prepareOrderIteratorAndMargins` sets the main axis margins on the renderer so that it can use them as part of the flex item's main axis size in `consturctFlexItem`. For single line flexboxes we could trim and set the bits beforehand for the first and last flex item, but with multi-line we don't have enough information until we start constructing the lines. So the current flexbox implementation has to do a couple of things:

1.) Modify the flex sizes while it is constructing a line inside `FlexibleBoxAlgorithm` by actively trimming the main axis margins
2.) Keep track of some state of which flex items are being trimmed and for which margins so that when the renderer goes through layout it can consult with flex layout via `shouldTrimChildMargin` so that it doesn't compute its margins when trimmed

Grid is a little more straightforward since the positioning is done early and we can know if a grid item needs its margin trimmed once the positioning phase is done. It's true that `setTrimmedMarginForChild` does have an unconditional call to `setMarginBeforeForChild`/`setMarginAfterForChild` that does not necessarily do anything in the case for grid since the renderer's margins have not been set.

We could extract out some of the necessary logic from `setTrimmedMarginForchild` into a lambda for the purposes of grid or perhaps make it a virtual function? Thoughts or preferences? 

With the introduction of these bits I think there's some cleanup/restructuring that I can do to make the logic a lot cleaner from the viewpoint of layout. I think the layout system (flex, grid, block) can determine if the renderer will have a margin trimmed and set the appropriate bits. Then `shouldTrimChildMargin` can be changed to just check for the existence of these bits. This would allow us to remove the extra state that was needed for flex layout. That will probably need to be done in a series of separate patches, though.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3815,8 +3817,11 @@ void RenderBox::computeAndSetBlockDirectionMargins(const RenderBlock& containing
-    if (containingBlock.shouldTrimChildMargin(marginSide, *this))
+    if (containingBlock.shouldTrimChildMargin(marginSide, *this)) {
+        if (isGridItem() && marginSide == MarginTrimType::BlockStart)
+            const_cast<RenderBox&>(*this).markMarginAsTrimmed(MarginTrimType::BlockStart);
```
please add a FIXME along the lines of, "This is supposed to be set when the trimmed margin is being...."

sammygill replied:
done!

sammygill committed:
[margin-trim] block-start margins for grid items in horizontal writing-mode should be reflected in computed style.
https://bugs.webkit.org/show_bug.cgi?id=253716
rdar://106559103

Reviewed by Alan Baradlay.

When the block-start margins for grid items that are adjacent to the
block-start edge of the grid, We need to set the rare data bits for
the item to indicate that the margin has been trimmed. We can then check
if this bit is set within ComputedStyleExtractor to get the trimmed
value for this margin by directly querying the renderer's m_marginBox.
In a horizontal writing mode this bit would correspond to the "top,"
margin value.

For block-start margins, the trimming occurs when a renderer goes
through layout and calls RenderBox::constrainBlockMarginInAvailableSpaceOrTrim,
which is called from RenderBox::computeBlockDirectionMargins. This
function consults the layout system, which is grid in this case, via
shouldTrimChildMargin to see if it should trim its margin. If this
returns true, then the renderer will not compute the margin and will
return 0_lu. We can use this opportunity to set the rare data for the
renderer to indicate that this margin has been trimmed.

Inside of ComputedStyleExtractor, we need to update
rendererCanHaveTrimmedMargin so that this returns true for grid items.
This will allow us to do 2 things:
1.) Have isLayoutDependent return true for grid items with margin-trim
set on the grid so that layout is forced when checking the "top," margin
for grid items
2.) Allow us to call box->hasTrimmedMargin when computing the margin
value (since that call is guarded by a call to rendererCanHaveTrimmedMargin)

Similarly, there are a couple of other extra pieces of code in RenderBox
(particularly in physicalToFlowRelativeDirectionMapping and hasTrimmedMargin)
that should now be used with grid items.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/grid-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/grid-block-start.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::rendererCanHaveTrimmedMargin):
(WebCore::isLayoutDependent):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::physicalToFlowRelativeDirectionMapping const):
(WebCore::RenderBox::hasTrimmedMargin const):
(WebCore::RenderBox::constrainBlockMarginInAvailableSpaceOrTrim const):

Canonical link: https://commits.webkit.org/262967@main
